crates/ruma-events/src/room/join_rules.rs

```diff
@@ -335,6 +341,53 @@ mod tests {
         );
     }
 
+    #[test]
+    fn invalid_allow_items() {
+        let json = r#"{
+            "join_rule": "restricted",
+            "allow": [
+                {
+                    "type": "m.room_membership",
+                    "room_id": "!mods:example.org"
+                },
+                {
+                    "type": "m.room_membership",
+                    "room_id": ""
+                },
+                {
+                    "type": "m.room_membership",
+                    "room_id": "not a room id"
+                },
+                {
+                    "type": "org.example.custom",
+                    "org.example.minimum_role": "developer"
+                },
+                {
+                    "not even close": "to being correct",
+                    "any object": "passes this test",
+                    "only non-objects in this array": "cause deserialization to fail"
+                }
+            ]
+        }"#;
+        let event: RoomJoinRulesEventContent = serde_json::from_str(json).unwrap();
+        match event.join_rule {
+            JoinRule::Restricted(restricted) => assert_eq!(
+                restricted.allow,
+                &[
+                    AllowRule::room_membership(owned_room_id!("!mods:example.org")),
+                    AllowRule::_Custom(Box::new(CustomAllowRule {
+                        rule_type: "org.example.custom".into(),
+                        extra: BTreeMap::from([(
+                            "org.example.minimum_role".into(),
+                            "developer".into()
+                        )])
+                    }))
+                ]
+            ),
+            rule => panic!("Deserialized to wrong variant: {rule:?}"),
+        }
```

Comment:
We use `assert_matches` for this, it will look like:

```rust
assert_matches!(event.join_rule, JoinRule::Restricted(restricted));
assert_eq!(
  restricted.allow,
  // …
);
```